File: implementations/react/src/patterns/structural/adapter/adapter.ts
import type { Adaptee } from "./adaptee";
import type { User } from "./interfaces/user";
import type { Target } from "./target";

/**
 * Clase adaptadora que convierte la interfaz de Adaptee a la interfaz Target.
 */

export class Adapter implements Target {

    private adaptee: Adaptee;

    constructor(adaptee: Adaptee) {
        this.adaptee = adaptee;
    }

    /**
     * Convierte la respuesta CSV de Adaptee a un array de objetos User.
     */
    async request(): Promise<User[]> {
        const csv = await this.adaptee.specificRequest();
        const lines = csv.trim().split(/\r?\n/);
        const rows = lines.slice(1).filter(Boolean);   

        return rows.map((row) => {
        const [id, name, email] = row.split(",");
        return { id: Number(id), name: name?.trim(), email: email?.trim() } as User;
        });
    }

}

File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/adapter/adapter.dart
import 'adaptee.dart';
import 'domain/user.dart';
import 'target.dart';

class Adapter implements Target {
  final Adaptee adaptee;
  Adapter(this.adaptee);

  @override
  Future<List<User>> request() async {
    final csv = await adaptee.specificRequest();
    final lines = csv.trim().split(RegExp(r'\r?\n'));
    final rows = lines.skip(1).where((e) => e.trim().isNotEmpty);
    return rows.map((row) {
      final parts = row.split(',');
      final id = int.tryParse(parts.elementAt(0)) ?? 0;
      final name = parts.length > 1 ? parts[1].trim() : '';
      final email = parts.length > 2 ? parts[2].trim() : '';
      return User(id: id, name: name, email: email);
    }).toList();
  }
}
